Added Windows support

Nitemice opened the pull request:
Added default cache path
Also had to replace call to grep with regex-based search in python,
because Windows doesn't have grep.

Nitemice committed:
Added Windows support

Added default cache path
Also had to replace call to grep with regex-based search in python,
because Windows doesn't have grep.

mikez commented on folders.py:
```diff
@@ -128,15 +129,22 @@ def get_folder(folder_id, data):
 
 def get_all_persistent_cache_files(path):
     """Get all files in PersistentCache storage with "start-group" marker."""
+    regex = ".*start-group.*"
```
Note: [`re.match`](https://docs.python.org/3/library/re.html#re.match) matches the beginning of the line, so the trailing `.*` can be omitted. Also, if you use [`re.search`](https://docs.python.org/3/library/re.html#re.search) you can skip the preceding `.*` as well.

Nitemice replied:
Good catch, thank you!
I wrote this code a while back but it wasn't working. So I had another look at it the other day, and the quick fix was to add the '.*' at each end. But I should have checked the docs for re.

mikez commented:
@Nitemice Thank you for taking the intiative! I'm sure that many Windows users can benefit from this.

I would like to clarify two things:

1. How common is the chosen Windows cache path? Will it work for 90% of the cases? See also discussion [here](https://github.com/mikez/spotify-folders/issues/6) which mentions two other paths.
2. How much of a speed-hit does the regex-based search cause? And, will the "cp437" encoding also work on non-Windows computers?

If there's a speed hit, I'd suggest to leave `get_all_persistent_cache_files` as-is and instead put your code into a Windows-version of the function, which gets run for Windows users. Another thing you can try, is to use the command as outlined [here](https://github.com/mikez/spotify-folders/issues/6#issuecomment-1115562629) and see if that speeds up things.

Nitemice commented:
Hi @mikez 

Thanks for building such a useful tool!

>     1. How common is the chosen Windows cache path? Will it work for 90% of the cases? See also discussion [here](https://github.com/mikez/spotify-folders/issues/6) which mentions two other paths.
>

To be honest, I wasn't totally sure if `C:\\Cache\\Spotify` was the correct path, but now that I've had a look at that discussion, I'm sure it's not.

My Spotify install is weird: it was originally installed stand-alone, then the Windows Store version was installed over the top. Plus, I have a custom cache location set, so that it doesn't take up room on my SSD. 

I think, given there are a few different ways to install Spotify, there will likely be a few "common" cache paths. That said, I checked with a family member and `C:\Users\USER\AppData\Local\Spotify\Storage` seems to be the path used by Spotify when installed directly (not from the Windows Store), so it's probably the most common and layman-friendly choice.
I will update the code to use that path by default.

>     2. How much of a speed-hit does the regex-based search cause? And, will the "cp437" encoding also work on non-Windows computers?
> 
> If there's a speed hit, I'd suggest to leave `get_all_persistent_cache_files` as-is and instead put your code into a Windows-version of the function, which gets run for Windows users. Another thing you can try, is to use the command as outlined [here](https://github.com/mikez/spotify-folders/issues/6#issuecomment-1115562629) and see if that speeds up things.

I don't think there's any speed impact. It seems to return output almost instantaneously.

I had another look at the encoding. I think it should actually be 'latin-1', and that should work on non-Windows platforms too. Basically, it just seems to need to be something that extends ASCII, and includes valid characters at a few specific code points that are used (probably as sentinels).

Personally, I like the idea of not depending on external tools, but if you feel it is slower, I'm willing to make your suggested changes.

I'll make those fixes and push it ASAP. If you can test it on a non-Windows machine, and tell me if it still runs satisfactorily fast, we can decide from there if I need to separate the Windows way from the others.

Nitemice committed:
Correct various details based on feedback

Corrected Windows cache path.
Corrected file encoding for cache file search.
Simplified regex pattern, and swapped to use 're.search'
instead of 're.match'.

mikez commented on folders.py:
```diff
@@ -128,15 +129,22 @@ def get_folder(folder_id, data):
     """Get all files in PersistentCache storage with "start-group" marker."""
+    regex = ".*start-group.*"
+    regObj = re.compile(regex)
+    res = []
```
res --> result (for consistency)

mikez commented on folders.py:
```diff
@@ -128,15 +129,22 @@ def get_folder(folder_id, data):
+            try:
+                with open(os.path.join(root, fname), encoding="cp437") as f:
+                    for line in f:
+                        if regObj.match(line):
```
Instead of opening as a text file, you can also open the stream in [binary mode](https://docs.python.org/3/library/functions.html#open). That way, no added decoding needs to be done.

`b'start-group' in line` suffices here. No need for regex. (Syntax assumes binary mode.)

mikez replied:
For readability, you could also extract some of the lines into a second helper function called `find(string, filepath)` which returns `True` or `False`.

mikez replied:
Lastly, if you see an easy way to make `PermissionError` backwards compatible to Python 2, that'll be great to. Right now, the code also works for Python 2. See also here: https://stackoverflow.com/a/18199529

mikez commented:
1. 👍
2. I tested it on my machine and it's fast. Let's use your version.

Also, added some minor comments to the code.

mikez commented:
@Nitemice What do you think?

Nitemice commented:
> @Nitemice What do you think?

I've pushed an update with some of the initial changes, but I'm still working on a solution for the `PermissionError` stuff.
The next commit I push will include basically all your suggestions.

Nitemice commented:
> @Nitemice What do you think?

I've pushed an update with some of the initial changes, but I'm still working on a solution for the `PermissionError` stuff.
The next commit I push will include basically all your suggestions.
It might take me a few days, as I'm busy with other stuff.

Nitemice committed:
Implement further feedback

Pulled out file search into a separate function
Swapped to catching generic exception, to be compatible with Python2
Also swapped to reading search files as binary to avoid encoding issues

mikez commented:
@Nitemice Except for the `WINDOWS_PERSISTENT_CACHE_PATH` fix, this looks wonderful to me, thank you. :)

mikez commented on folders.py:
```diff
@@ -24,9 +23,12 @@
 LINUX_PERSISTENT_CACHE_PATH = os.path.join(
     os.getenv('XDG_CACHE_HOME', '~/.cache'), 'spotify/Storage')
+WINDOWS_PERSISTENT_CACHE_PATH = os.path.join(
+    os.getenv('LOCALAPPDATA'), 'Spotify\\Storage')
```
@Nitemice You may want to have a fallback in `os.getenv` if `LOCALAPPDATA` is not found. Otherwise non-Windows computers will return an error here, as it tries to do `os.path.join(None, ...)`.

Nitemice replied:
I don't suppose you have any suggestions for an appropriate default path. I can't think of anything that's going to possibly be a valid path that doesn't use an environment variable.

Also, I was able to work out the path used by Spotify when it's installed from the Windows Store, so if there's a way to detect if a path exists, we could detect which to use. It's `%LOCALAPPDATA%\Packages\SpotifyAB.SpotifyMusic_zpdnekdrzrea0\LocalState\Spotify\Storage`.

mikez replied:
How about rewriting those lines to

```
if sys.platform == "darwin":
    # macOS
    PERSISTENT_CACHE_PATH = (
        "~/Library/Application Support/Spotify/PersistentCache/Storage"
    )
elif sys.platform == "win32":
    # Windows
    PERSISTENT_CACHE_PATH = os.path.join(
        os.getenv("LOCALAPPDATA"), "Spotify", "Storage"
    )
else:
    # Linux
    PERSISTENT_CACHE_PATH = os.path.join(
        os.getenv("XDG_CACHE_HOME", "~/.cache"), "spotify/Storage"
    )
```

If Windows is more involved, you could also write a `get_windows_persistent_cache_path()` helper function and use `os.path.exists`.

That said, maybe we can first commit what we have now, and then do the more involved Windows solution in an upcoming pull request? What do you think?

Nitemice replied:
I actually wrote some very similar code after I wrote my comment last night. If you think it's a good idea, I'm happy to finish it off.
But yeah, fair enough. Let's finish this stuff first before going down that path. In which case, what is there left to do before we merge this? Just add a default for `os.getenv('LOCALAPPDATA')`?

mikez replied:
If we do the `if-else` route like sketched above, I think no default value is needed. I think we can assume that Windows will have the `LOCALAPPDATA` variable set. The earlier suggestion was for the other platforms to not error, but with the if-else that's no longer needed.

Nitemice replied:
So should I go ahead with the `if-else` stuff, or do we want to merge the code as is, or is there something else that need to be done first?

mikez replied:
`if-else` is needed, so the code doesn't break on Linux or MacOS systems. Otherwise, I think we're good. :)

Nitemice committed:
Refactor persistent cache path logic

Persistent cache path can now detect two different Windows install
locations, and will only check them if on Windows.
